lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -526,6 +526,36 @@ void ESIPortsPass::updateInstance(RTLExternModuleOp mod, InstanceOp inst) {
   inst.erase();
 }
 
+namespace {
+/// Eliminate back-to-back wrap-unwraps to reduce the number of ESI channels.
+struct RemoveWrapUnwrap : public OpConversionPattern<WrapValidReady> {
+public:
+  RemoveWrapUnwrap(MLIRContext *ctxt) : OpConversionPattern(ctxt) {}
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult match(Operation *op) const final {
+    auto wrap = dyn_cast<WrapValidReady>(op);
+    if (!wrap)
+      return failure();
+    if (!wrap.chanOutput().hasOneUse())
+      return failure();
+    if (!isa<UnwrapValidReady>(wrap.chanOutput().use_begin()->getOwner()))
```

Comment:
It might be simpler to make this pattern a single matchAndWrite, and replace the `isa` here with the `dyn_cast` below.